### PR TITLE
[RemoteMirrors] Rewrite the MachO reader path.

### PR DESCRIPTION
**Pull request by dcci**
There are several problems fixed by this commit.
-> It removes the dependency on a system header.
-> It works when the pointer size for host and target don't match (i.e. 32-bits/64-bits and viceversa)
-> It reads the file lazily, mapping only the sections that are needed. The old approach mapped the whole __TEXT segment, causing significant slowdowns when trying to integrate this API in the debugger. Now it's almost instant (10 seconds vs milliseconds).
    
<rdar://problem/42306551>

### CODE REVIEW

**Comment by dcci** [comment]
@swift-ci please test

File: include/swift/Reflection/ReflectionContext.h [hunk 1]
```diff
@@ -112,83 +103,119 @@ class ReflectionContext
...
-        {{TyperefMd.first.startAddress(), TyperefMd.first.endAddress()}, 0},
-        {{ReflStrMd.first.startAddress(), ReflStrMd.first.endAddress()}, 0},
+        {{FieldMdSec.first.first, FieldMdSec.first.second}, 0},
+        {{AssocTySec.first.first, AssocTySec.first.second}, 0},
+        {{BuiltinTySec.first.first, BuiltinTySec.first.second}, 0},
+        {{CaptureSec.first.first, CaptureSec.first.second}, 0},
+        {{TypeRefMdSec.first.first, TypeRefMdSec.first.second}, 0},
+        {{ReflStrMdSec.first.first, ReflStrMdSec.first.second}, 0},
         LocalStartAddress,
         RemoteStartAddress};
```
Comment (mikeash) [inline include/swift/Reflection/ReflectionContext.h]: I'm not quite sure that `LocalStartAddress` and `RemoteStartAddress` make sense anymore. My (limited) understanding is that they are used to translate between an address in the local space and an address in the remote space, which assumes that all of the reflection sections have the same delta between the two address spaces. That was true when the entire TEXT segment was being mapped in one shot, but is not necessarily true with these changes.

It looks like the only code that actually uses those addresses is `getRemoteAddrOfTypeRefPointer` in `TypeRefBuilder.cpp`. It only translates addresses within the `TypeReference` sections. Maybe we can get what we need by just storing the remote address of that one section.

**Comment by dcci** [comment]
Whilst this passes the regular test suite, it fails with validation-test (probably uncovering some bugs in the implementation). I'm going to take a look at it next.

**Comment by dcci** [comment]
@swift-ci please test

File: include/swift/Reflection/ReflectionContext.h (hunk 1 above)
Comment (dcci) [reply include/swift/Reflection/ReflectionContext.h]: I spoke with Nick and he recommended to just map the whole `__TEXT` as taking in account the delta between sections may be complicated. This is what I did. If this turns out to be a bottleneck we might reconsider it but I really want to make some progress on the debugger integration.

**Comment by dcci** [comment]
@swift-ci please test

**Comment by dcci** [comment]
pretty sure this is not my fault:
```
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/tests/Output/swift-build-self-host.py.tmp.dir/swiftpm/Sources/SPMLLBuild/llbuild.swift:15:19: error: no such module 'llbuild'
18:39:47 @_exported import llbuild
18:39:47                   ^
18:39:47 error: terminated(1): /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/usr/bin/swift-build-tool -f /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/none-swift_package_sandbox_linux-x86_64/tests/Output/swift-build-self-host.py.tmp.dir/swiftpm/.build/debug.yaml main output:
```

@ddunbar is this a known issue?

**Comment by dcci** [comment]
@swift-ci please smoke test linux

**Comment by dcci** [comment]
@swift-ci please smoke test linux

File: include/swift/Reflection/ReflectionContext.h (hunk 1 above)
Comment (mikeash) [reply include/swift/Reflection/ReflectionContext.h]: Makes sense to me. Is it actually slower to map everything in one chunk compared to mapping several smaller chunks? I'd have expected each mapping operation to be approximately constant time regardless of size, but I could certainly be wrong.

**Comment by ddunbar** [comment]
@aciidb0mb3r can you take a look at the llbuild thing?

**Comment by aciidb0mb3r** [comment]
Should be resolved as of yesterday night.

**Comment by aciidb0mb3r** [comment]
@swift-ci please test linux

**Comment by dcci** [comment]
@swift-ci test and merge

**Comment by dcci** [comment]
@swift-ci test and merge

**Comment by dcci** [comment]
@swift-ci please smoke test and merge

**Comment by dcci** [comment]
@swift-ci please smoke test and merge

**Comment by clackary** [comment]
@swift-ci smoke test and merge

**Pull request merged**